### ISSUE TITLE
how to display edge detection result from the camera data?

### ISSUE THREAD

**Issue by chankim** [carla-simulator/carla#2210]
I'm trying to display a edge detection on Carla (http://carla.org/) client. I'm using versino 0.9.6.
In the example client script automatic_control.py,
I modified(added edge detection image and display it it using opencv) in CameraManager's parse_image(weak_self, image) function as below.(for case when it's not lidar).

    image.convert(self.sensors[self.index][1])
    array = np.frombuffer(image.raw_data, dtype=np.dtype("uint8"))
    array_1d = np.copy(array)
    array_4d = array_1d.reshape((720,1280,4))
    #print('type(array_4d) = ', type(array_4d))
    #print('shape = ', array_4d.shape)
    gray = cv2.cvtColor(array_4d, cv2.COLOR_RGB2GRAY)
    blur = cv2.GaussianBlur(gray, (5,5), 0)
    canny = cv2.Canny(blur, 50, 150)
    cv2.imshow("result", canny)
    cv2.waitKey(1)
    array = np.reshape(array, (image.height, image.width, 4))
    array = array[:, :, :3]
    array = array[:, :, ::-1]
    self.surface = pygame.surfarray.make_surface(array.swapaxes(0, 1))

But the edge detection result is display only once and not updated. But the pygame HUD(head-up-display) window shows the normal RGB camera images as the simulation goes on. I tried using synchronous mode with fixed_delta_seconds setting in vain. What should I do?

I have posted the same question on stackoverflow : https://stackoverflow.com/questions/58703741/opencv-python-imshow-not-updated-in-a-loop-even-with-waitkey1

**Comment by chankim** [comment]
so my question is how to dispay an additive image on opencv or pyplot. I tried both but either the main pygame window or opencv/pyplot window is not updated. I tried putting the display function in the main game loop but it's the same. Hope someone could give me a guide.

**Comment by chankim** [comment]
I changed the automatic_control.py so that displaying the edge detection is done in the game_loop(). It's like this (cropped_image is made inside the camera manager). The main display is not updated and the server FPS is 23 but client fps is 0.  and the "loop1 ..." is printed but "loop2 ..." is never printed. What's wrong in here? Please someone help me here ..

        while True:
            if controller.parse_events(client, world, clock):
                return

            # as soon as the server is ready continue!
            world.world.wait_for_tick(1.0)

            world.tick(clock)
            world.render(display)
            pygame.display.flip()
            control = agent.run_step()
            control.manual_gear_shift = False
            print('loop1 ...')
            plt.imshow(world.camera_manager.cropped_image)
            plt.show(1)
            print('loop2 ...')
            world.player.apply_control(control)

**Comment by chankim** [comment]
Turns out using pyplot together with the pygame display doesn't work well(I don't know why).
So I used the pygame display to show it.
I've managed to display the lane detection result on the original image. I made a new overlay image and made the pygame Surface display the overlay image. The arrayx below is the overlay image.
           array = np.reshape(arrayx, (image.height, image.width, 4))
            array = array[:, :, :3]
            array = array[:, :, ::-1]
            self.surface = pygame.surfarray.make_surface(array.swapaxes(0, 1))
![lane_overlay](https://user-images.githubusercontent.com/824480/68929965-bd600500-07d0-11ea-88b6-9914a14d36e4.JPG)
